### PR TITLE
cosine_similarity and also slot_similarity_tuple_matcher

### PR DESCRIPTION
**Pull request by bsharpataz**

### CODE REVIEW

File: deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                                            # tuple2 shape: (batch size, num_slots, embed_dimension)
+        # Check that the tuples have the same number of slots.
+        assert K.int_shape(tuple1_input)[1] == K.int_shape(tuple2_input)[1]
+
+        # Calculate the cosine similarities.
+        # shape: (batch size, num_slots)
+        similarities = self.similarity_function.compute_similarity(tuple1_input, tuple2_input)
+
+        # Remove any similarities if one of the corresponding slots was all padding.
+        # TODO(becky): need this to take a mask for each input of None or shape: (batch size, num_slots)
```
Comment (matt-gardner) [inline deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py]: I think this TODO can be removed?  Or is there something else that needs to be done?

File: deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        The activation of the NN output layer
+
+    Notes
+    _____
+    This layer is incompatible with the WordsAndCharacters tokenizer.
+    """
+    def __init__(self, similarity_function_choice, num_hidden_layers: int=1, hidden_layer_width: int=4,
+                 initialization: str='glorot_uniform', hidden_layer_activation: str='tanh',
+                 final_activation: str='sigmoid', **kwargs):
+        self.input_dim = None
```
Comment (matt-gardner) [inline deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py]: I don't think this is used anywhere.

File: deep_qa/layers/tuple_matchers/encoded_tuple_matcher.py [hunk 3]
```diff
@@ -28,7 +29,7 @@ def __init__(self, text_trainer, tuple_matcher_params: Dict[str, Any]):
         if tuple_matcher_params is None:
             tuple_matcher_params = {}
         tuple_matcher_choice = get_choice_with_default(tuple_matcher_params,
-                                                       "type",
+                                                       "encoded_match_type",
```
Comment (matt-gardner) [inline deep_qa/layers/tuple_matchers/encoded_tuple_matcher.py]: Oh, nevermind.  I thought there was some nested structure here.  This is probably fine.  ("probably" because I was wondering if we should make this structure nested...  But that can wait for another PR, if we decide we want to change it.)

File: tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        num_slots = 3
+        embed_dimension = 4
+        self.tuple1_input = Input(shape=(num_slots, embed_dimension), dtype='float32', name="input_tuple1")
+        self.tuple2_input = Input(shape=(num_slots, embed_dimension), dtype='float32', name="input_tuple2")
+        self.num_hidden_layers = 1
+        self.hidden_layer_width = 2
+        self.hidden_layer_activation = 'linear'
+
+        # tuple1 shape: (batch size, num_slots, embed_dimension)
+        self.tuple1 = numpy.asarray([[numpy.random.random(embed_dimension) for _ in range(num_slots)]],
```
Comment (matt-gardner) [inline tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py]: You can just do `self.tuple1 = numpy.random.rand(1, num_slots, embed_dimension)`

File: tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        # Testing general unmasked case.
+        similarity_function = CosineSimilarity(name="cosine_similarity")
+        cosine_similarities = similarity_function.compute_similarity(K.variable(self.tuple1), K.variable(self.tuple2))
+
+        # Desired_overlap gets fed into the inner NN.
+        dense1_activation = numpy.dot(K.eval(cosine_similarities), dense_hidden_weights)
+        final_score = numpy.dot(dense1_activation, score_weights)
+        desired_result = logistic.cdf(final_score)
+        print(desired_result)
```
Comment (matt-gardner) [inline tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py]: You can remove this print statement.

File: tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        self.tuple2[:, 1, :] = numpy.zeros(4)
+        match_layer = SlotSimilarityTupleMatcher("cosine_similarity",
+                                                 self.num_hidden_layers,
+                                                 self.hidden_layer_width,
+                                                 hidden_layer_activation=self.hidden_layer_activation)
+        output = match_layer([masked_tuple1, masked_tuple2])
+        mask_model = Model([self.tuple1_input, self.tuple2_input], output)
+
+        # Testing masked case
+        # Testing general unmasked case.
```
Comment (matt-gardner) [inline tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py]: Comments here are contradictory.

File: tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        # Get the initial weights for use in testing
+        dense_hidden_weights = K.eval(mask_model.trainable_weights[0])
+        score_weights = K.eval(mask_model.trainable_weights[1])
+
+        # Desired_overlap gets fed into the inner NN.
+        dense1_activation = numpy.dot(K.eval(masked_cosine_similarities), dense_hidden_weights)
+        final_score = numpy.dot(dense1_activation, score_weights)
+        desired_result = logistic.cdf(final_score)
+        result = mask_model.predict([self.tuple1, self.tuple2])
+        assert_array_almost_equal(result, desired_result)
```
Comment (matt-gardner) [inline tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py]: This red icon at the end of the line means that your file doesn't end in a newline.  Can you fix that?

File: tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        score_weights = K.eval(model.trainable_weights[1])
+
+        # Testing general unmasked case.
+        similarity_function = CosineSimilarity(name="cosine_similarity")
+        cosine_similarities = similarity_function.compute_similarity(K.variable(self.tuple1), K.variable(self.tuple2))
+
+        # Desired_overlap gets fed into the inner NN.
+        dense1_activation = numpy.dot(K.eval(cosine_similarities), dense_hidden_weights)
+        final_score = numpy.dot(dense1_activation, score_weights)
+        desired_result = logistic.cdf(final_score)
```
Comment (matt-gardner) [inline tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py]: What's this cdf here for?

File: deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        - tuple_2_input (the background_tuple), shape ``(batch size, num_slots, embed_dimensions)``,
+          and again, there needs to be a corresponding mask of shape (batch size, num_slots) (or None)
+          that indicates whether a given slot was all padding.
+
+    Output:
+        - entailment score, shape ``(batch, 1)``
+
+    Parameters
+    ----------
+    - similarity_function_choice: str
```
Comment (bsharpataz) [inline deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py]: I thought that the Layers were supposed to have not dict params for the __init__...?  I originally had that for WordSimilarityTupleMatcher and you had asked me to change it to use kwargs instead (for loading?).  I am happy to match whatever style is needed though, just let me know/confirm and I'll make the change.

File: deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+    - hidden_layer_activation : string, default='relu'
+        The activation of the NN hidden layers
+
+    - final_activation : string, default='sigmoid'
+        The activation of the NN output layer
+
+    Notes
+    _____
+    This layer is incompatible with the WordsAndCharacters tokenizer.
```
Comment (bsharpataz) [inline deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py]: sorry - you're right, it's a copy-paste bug! thanks!

File: tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py (hunk 8 above)
Comment (bsharpataz) [inline tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py]: because the output node has a sigmoid activation

File: deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py (hunk 9 above)
Comment (bsharpataz) [inline deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py]: oh -- shoot, you meant that I need to pass in a dict for the *similarity_function*
ok -- let me work on that

**Comment by bsharpataz** [comment]
sure can :)

File: deep_qa/layers/wrappers/encoder_wrapper.py [hunk 11]
```diff
@@ -23,3 +23,4 @@ def compute_mask(self, x, input_mask=None):
         else:
             # An output bit is 0 only if the  bits corresponding to all input words are 0.
             return K.any(input_mask, axis=-1)
+
```
Comment (matt-gardner) [inline deep_qa/layers/wrappers/encoder_wrapper.py]: You can just get `git checkout master deep_qa/layers/wrappers/encoder_wrapper.py` to revert this.  No need to have this here.

File: example_experiments/simple_experiment.json [hunk 12]
```diff
@@ -6,6 +6,10 @@
             "type": "bow"
         }
     },
+    "pretrained_embeddings_file": "filename to gzipped glove vectors get from efs/data/dlfa/glove.840B.300d.txt.gz",
```
Comment (matt-gardner) [inline example_experiments/simple_experiment.json]: I wouldn't check this in with a fake path, or with the comment below.  If you want the simple experiment to have these params as an example (which I think is fine), it should be with params that actually work if you run this from a machine with the EFS drive.  So just change this to the correct path.

File: example_experiments/svo_mc_tuple_entailment.json [hunk 13]
```diff
@@ -15,17 +15,25 @@
...
+        }
+    },
     "noisy_or_param_init": "uniform",
     "num_question_tuples": 30,
     "num_background_tuples": 50,
     "num_tuple_slots": 4,
     "word_sequence_length": 5,
     "num_answer_options": 4,
-    "train_files": ["/Users/rebeccas/data/brittleness/omni4_train_brittleness.tuples.labeled"]
+    //"train_files": ["/Users/rebeccas/data/brittleness/omni4_train_brittleness.tuples.labeled"]
```
Comment (matt-gardner) [inline example_experiments/svo_mc_tuple_entailment.json]: I probably wouldn't check in this change.

File: tests/layers/wrappers/time_distributed_test.py [hunk 14]
```diff
@@ -6,6 +6,7 @@
 from keras.layers import Input, Lambda
 from keras.models import Model
 
+
```
Comment (matt-gardner) [inline tests/layers/wrappers/time_distributed_test.py]: This newline is not needed.

File: tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+        masked_cosine_similarities = cosine_similarities * mask2
+
+        # Get the initial weights for use in testing
+        dense_hidden_weights = K.eval(mask_model.trainable_weights[0])
+        score_weights = K.eval(mask_model.trainable_weights[1])
+
+        # Desired_overlap gets fed into the inner NN.
+        dense1_activation = numpy.dot(K.eval(masked_cosine_similarities), dense_hidden_weights)
+        final_score = numpy.dot(dense1_activation, score_weights)
+        desired_result = logistic.cdf(final_score)
```
Comment (matt-gardner) [inline tests/layers/tuple_matchers/slot_similarity_tuple_matcher_test.py]: Comment on this `logistic.cdf` call would also be nice.

File: deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+        super(SlotSimilarityTupleMatcher, self).__init__(**kwargs)
+        sim_function_choice = get_choice_with_default(similarity_function,
+                                                      'type',
+                                                      list(similarity_functions.keys()))
+        similarity_function['name'] = self.name + '_similarity_function'
+        self.similarity_function = similarity_functions[sim_function_choice](**similarity_function)
+
+    def get_config(self):
+        base_config = super(SlotSimilarityTupleMatcher, self).get_config()
+        config = {'similarity_function_choice': self.similarity_function_choice,
```
Comment (matt-gardner) [inline deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py]: This isn't correct - copy what `MatrixAttention` does here.

File: deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+        # Parameters for the shallow neural network
+        self.num_hidden_layers = num_hidden_layers
+        self.hidden_layer_width = hidden_layer_width
+        self.hidden_layer_init = initialization
+        self.hidden_layer_activation = hidden_layer_activation
+        self.final_activation = final_activation
+        self.hidden_layer_weights = []
+        self.score_layer = None
+        super(SlotSimilarityTupleMatcher, self).__init__(**kwargs)
+        sim_function_choice = get_choice_with_default(similarity_function,
```
Comment (matt-gardner) [inline deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py]: You need to copy what `MatrixAttention` does basically exactly.  In particular, you need a line like `if similarity_function is None: similarity_function = {}`.

File: deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py (hunk 16 above)
Comment (bsharpataz) [inline deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py]: ah, yeah, forgot to change that when I changed to a dict, good catch

File: example_experiments/simple_experiment.json [hunk 18]
```diff
@@ -6,7 +6,7 @@
             "type": "bow"
         }
     },
-    "pretrained_embeddings_file": "filename to gzipped glove vectors get from efs/data/dlfa/glove.840B.300d.txt.gz",
+    "pretrained_embeddings_file": "/efs/data/dlfa/glove/glove.840B.300d.txt.gz",
     "project_embeddings": false,
     "fine_tune_embeddings": false,
     // check the docs for more info
```
Comment (matt-gardner) [inline example_experiments/simple_experiment.json]: Remove this comment, too?

File: deep_qa/layers/backend/squeeze.py [hunk 19]
```diff
@@ -26,6 +26,11 @@ def compute_mask(self, inputs, mask=None):
         # pylint: disable=unused-argument
         if mask is None:
             return None
+        # If the mask doesn't need squeezing, just return it.
+        # TODO(becky): better to check if the last mask dim == 1?
+        if K.ndim(inputs) == K.ndim(mask) + 1:
```
Comment (matt-gardner) [inline deep_qa/layers/backend/squeeze.py]: Where is this necessary?

Comment (bsharpataz) [inline deep_qa/layers/backend/squeeze.py]: i need it for the tuple_inference: https://github.com/allenai/deep_qa/blob/master/deep_qa/models/multiple_choice_qa/tuple_inference.py#L163-L164
At least when I was using the EncoderWrappers (not TimeDistributedWithMask). the resulting mask at the end of the matching was one dimension shorter and so didn't need squeezing (and using squeeze made it crash bc of Squeeze.compute_mask).  That said, if your change to TimeDistributedWithMask makes the dimensions of the tensor and its mask identical (either way, both with the extra dim of size 1 or both without it) then it wouldn't be necessary.

Comment (matt-gardner) [inline deep_qa/layers/backend/squeeze.py]: I'm pretty sure the shapes work out now that this isn't needed.  Can you remove it and see if it still works?

**Comment by bsharpataz** [comment]
I can't until I have your changes to TimeDistributedWithMask.  As it is
now, I don't think it builds.  If we do the merges, I can remove it and see
if it works, and if so I'll remove it.  Does that work?

**Comment by matt-gardner** [comment]
I just checked out the code to see if tests pass if I remove what you added to `Squeeze`, and they do.  I would recommend removing it and adding it to the next PR, if it's still necessary.

**Comment by bsharpataz** [comment]
yeah, the tests only test the model not crashing with the
WordOverlapTupleMatcher, duh.  so tests pass.
ok, will remove
